Привет, меня зовут Люман Аблаев. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

### <font color='orange'>Общее впечатление</font>
* Большое спасибо за проделанную работу!
* Каждый шаг выполнен уверенно, видно, что приложено много усилий.
* Отличное использование функций. Ты освоил данный навык, довольно, хорошо!
* Выводы и рассуждения получились содержательными, их было занимательно читать. Я оставил небольшой совет по markdown стилю.
* Я оставил немного советов, надеюсь ты обратишь на них внимание и они будут полезными.
* Эта работа зачтена. Удачи в дальнейшем обучении.

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-block alert-success">
<b>Успех:</b> Спасибо за описание проекта.
</div>

### Шаг 1. Откройте файл с данными и изучите общую информацию. 


<div class="alert alert-block alert-info">
    <b>Совет:</b> Лучше собирать все импорты в начале проекта. Когда пользователь будет запускать твой проект целиком. Лучше, чтобы при отсуствии какого-либо пакета, код упал в начале чем где-то посреди. Иногда код может выполнятся очень долгое время, и будет неприятно, если он упадет из-за такой мелочи =)
</div>

In [1]:
import pandas as pd
df = pd.read_csv('/datasets/data.csv')
df.head(50)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   
30         1  335581.668515         62              среднее             1   
31         0   -1682.083438         47            начальное             3   
32         0   -4649.910832         34              СРЕДНЕЕ             1   
33         0   -1548.637544         48              среднее             1   
34         0   -4488.067031         35              среднее             1   
35         0  394021.072184         68              среднее             1   
36         0    -176.216688         33              среднее             1   
37         0   -6448.810860         43               ВЫСШЕЕ             0   
38         0    -597.881827         25               высшее             0   
39         0    -650.587796         31               высшее             0   
40         0   -1030.362606         30              среднее             1   
41         0            NaN         50              среднее             1   
42         0   -1257.496190         20  неоконченное высшее             2   
43         0   -4375.681384         43  неоконченное высшее             2   
44         1   -1362.041728         26              СРЕДНЕЕ             1   
45         1   -1039.451741         49              среднее             1   
46         2   -2262.712304         37              среднее             1   
47         1   -2689.137274         33               высшее             0   
48         0   -3341.067886         45              среднее             1   
49         0   -1181.443228         54              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0  

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### Вывод

Данные изучены. Явными проблемами, мешающими обработке данных являются: 
1. Пропуски NaN в столбце "days_employed" и "total_income"
2. Отсутствие единого регистра в колонке "education"
3. Вариантивность обозначения цели займа в колонке "purpose"
4. Отрицательные значения в столбцах "days_employed" и "children"

<div class="alert alert-block alert-success">
<b>Успех:</b>  Загрузка данных и первичный анализ сделаны верно.
</div>

### Шаг 2. Предобработка данных

### Обработка пропусков

In [3]:
print(df.isnull().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


In [4]:
df['days_employed'] = abs(df['days_employed'])
df['children'] = abs(df['children'])

In [5]:
days_employed_median = df['days_employed'].median()

In [6]:
total_income_median = df['total_income'].median()

In [7]:
days_employed_median

2194.220566878695

In [8]:
total_income_median

145017.93753253992

In [9]:
df['days_employed'] = df['days_employed'].fillna(days_employed_median)
df['total_income'] = df['total_income'].fillna(total_income_median)

In [10]:
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              Среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              СРЕДНЕЕ             1   
8          2    6929.865299         35               ВЫСШЕЕ             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0    2194.220567         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12  145017.937533                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

### Вывод

Изначально определим количество пропусков и их местоположение, используя метод isnull().
Становится ясно, что пропуски есть только в столбцах "days_employed" и "total_income". 
Количество пропусков одинаковое - 2174. Детально изучив таблицу, становиятся ясно, что отсутствие информации в ячейке по столбцу "days_employed", это признак отсутствия информации в той же ячейке по столбцу "total_income". Вероятно, клиент не предоставил эту информацию банку либо сумма кредита не требовала предоставления этой информации.

Количество пропусков в датафрейме - 2174, всего строк датафрейма - 21525, значит пропусков 10%. Для корректной работы с данными, пропуски в ячейках нужно убрать. Удалить пропуски не получится - их слишком много, поставить нули - исказит реальную картину, заменить пропуски средним значением по столбцу - при учете выдающихся значений, данные будут не корректными. На мой взгляд оптимальным методом для выведения среднего по столбцу является медиана, ее и будем использовать при замене пропусков.

Столбцы "children" и "days_employed" приведем к неотрицательному значению (модулю) путем метода abs().

<div class="alert alert-block alert-success">
<b>Успех:</b>  Хороший выбор заполнения пропусков - медианной, она устойчива к выбросам. Однако есть более продвинутный подход, заполнять пропуски опираясь только на медианные значений из соответсвенных групп других колонок. Например для пенсионеров - медианна среди пенсонеров, для бизнессменов - медиана среди бизнесменов.
</div>

### Замена типа данных

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [12]:
df['days_employed'] = df['days_employed'].astype('int64')
df['total_income'] = df['total_income'].astype('int64')


In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


### Вывод

Воспользовавшись методом info(), определяем, что вещественный тип данных встречается в столбцах "days_employed" и "total_income". Меняем float64 на int64 с помощью метода astype('int64'), тк перевести нужно в целочисленное значение.

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, твоя конвертация  типов  не только улучшает визуальные восприятие данных, но также экономит память и время.
</div>

### Обработка дубликатов

In [14]:
df['education'] = df['education'].str.lower()
df.duplicated().sum()

71

In [15]:
df = df.drop_duplicates().reset_index(drop=True)

In [16]:
df.duplicated().sum()

0

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 12 columns):
children            21454 non-null int64
days_employed       21454 non-null int64
dob_years           21454 non-null int64
education           21454 non-null object
education_id        21454 non-null int64
family_status       21454 non-null object
family_status_id    21454 non-null int64
gender              21454 non-null object
income_type         21454 non-null object
debt                21454 non-null int64
total_income        21454 non-null int64
purpose             21454 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


### Вывод

Определяем количество полных дубликатов. Нас не интересуют дубликаты по каждому столбцу, нас интересует явное дублирование строки при полном повторении всех значений каждого столбца. Используем метод duplicated().sum() для определения количества всех дубликатов и их полное удаление cо сбросом индексов помощью метода drop_duplicates().reset_index(drop=True)

Переведем в нижний регистр столбец "education" c помощью метода lower() - исключим дубликаты с учетом регистра.

На мой взгляд, появление полных дубликатов возможно, путем соединения нескольких таблиц от разных сотрудников в одну таблицу, где определенная строка присутствовала как в первой таблице, так и во второй. К примеру, над определенным клиентом работали 2 сотрудника банка.

<div class="alert alert-block alert-success">
<b>Успех:</b> Молодец, что привел колонку education к нижнему регистру.
</div>

### Лемматизация

In [18]:
text = df['purpose'].unique()

In [19]:
text

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [20]:
#from pymystem3 import Mystem
#m = Mystem() 
#lemmas = m.lemmatize(text)
#print(lemmas) 
# AttributeError: 'numpy.ndarray' object has no attribute 'splitlines'
from nltk.stem import SnowballStemmer 
russian_stemmer = SnowballStemmer('russian')  
for word in text:
    print ('Исходное слово - ' + word + ', после стемминга - ' + russian_stemmer.stem(word)) 

Исходное слово - покупка жилья, после стемминга - покупка жил
Исходное слово - приобретение автомобиля, после стемминга - приобретение автомобил
Исходное слово - дополнительное образование, после стемминга - дополнительное образован
Исходное слово - сыграть свадьбу, после стемминга - сыграть свадьб
Исходное слово - операции с жильем, после стемминга - операции с жил
Исходное слово - образование, после стемминга - образован
Исходное слово - на проведение свадьбы, после стемминга - на проведение свадьб
Исходное слово - покупка жилья для семьи, после стемминга - покупка жилья для сем
Исходное слово - покупка недвижимости, после стемминга - покупка недвижим
Исходное слово - покупка коммерческой недвижимости, после стемминга - покупка коммерческой недвижим
Исходное слово - покупка жилой недвижимости, после стемминга - покупка жилой недвижим
Исходное слово - строительство собственной недвижимости, после стемминга - строительство собственной недвижим
Исходное слово - недвижимость, после стемм

In [21]:
house = []
auto = []
education = []
wedding = []
for query in text:
    for word in query.split():
        stemmed_word = russian_stemmer.stem(word)
        if stemmed_word == 'жил' or stemmed_word == 'недвижим':
            house.append(query) 
        if stemmed_word == 'автомобил':
            auto.append(query)
        if stemmed_word == 'образован':
            education.append(query)
        if stemmed_word == 'свадьб':
            wedding.append(query)

print(house) 
print()
print(auto)
print()
print(education)
print()
print(wedding)

['покупка жилья', 'операции с жильем', 'покупка жилья для семьи', 'покупка недвижимости', 'покупка коммерческой недвижимости', 'покупка жилой недвижимости', 'покупка жилой недвижимости', 'строительство собственной недвижимости', 'недвижимость', 'строительство недвижимости', 'операции с коммерческой недвижимостью', 'строительство жилой недвижимости', 'строительство жилой недвижимости', 'жилье', 'операции со своей недвижимостью', 'покупка своего жилья', 'операции с недвижимостью', 'покупка жилья для сдачи', 'ремонт жилью']

['приобретение автомобиля', 'на покупку подержанного автомобиля', 'на покупку своего автомобиля', 'сделка с подержанным автомобилем', 'автомобиль', 'свой автомобиль', 'сделка с автомобилем', 'на покупку автомобиля']

['дополнительное образование', 'образование', 'заняться образованием', 'получение образования', 'получение дополнительного образования', 'получение высшего образования', 'профильное образование', 'высшее образование', 'заняться высшим образованием']

['сы

### Вывод

Использовал Стемминг, тк Леммингом воспользоваться не смог из-за ошибки : AttributeError: 'numpy.ndarray' object has no attribute 'splitlines'.
Изначально через метод unique() определил все уникальные значения. Затем с помощью библиатеки NLTK определил изначальные форма целей кредита в столбце 'purpose'. Стало ясно, что основных целей кредита 4: автомобиль, образование, недвижимость и жилье (объеденил в одно), свадьба. Эти 4 цели понадобятся для категоризации данных.


<div class="alert alert-block alert-success">
<b>Успех:</b>  Стемминг применен корректно. Получились верные группы. Насчет лемматизации, скорее всего ошибка в том, что тебе нужно было проходиться циклом по значениям колонки purpose и применять lemmatize к ним.
</div>

### Категоризация данных

In [22]:
df['purpose'].value_counts()

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
жилье                                     646
покупка жилья                             646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

In [23]:
def purpose_def(word):
    if 'жил' in word:
        return 'недвижимость'
    if 'недвижим' in word:
        return 'недвижимость'
    if 'автомобил' in word:
        return 'автомобиль'
    if 'образован' in word:
        return 'образование'
    if 'свадьб' in word:
        return 'свадьба'

In [24]:
purpose_def('строительство жилой недвижимости')

'недвижимость'

In [25]:
df['purpose_group'] = df['purpose'].apply(purpose_def)

In [26]:
def total_income_def(income):
    if income < 75000:
        return 'низкий доход'
    if income < 150000:
        return 'средний доход'
    if income >= 150000:
        return 'высокий доход'

In [27]:
df['total_income_group'] = df['total_income'].apply(total_income_def)

<div class="alert alert-block alert-info">
<b>Совет:</b> Интересно, почему ты решил разделить группы именно на такие разбиения, и получились ли группы равномерно распределенные, чтобы их можно было объективно сравнивать.
</div>

In [28]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose purpose_group total_income_group  
0               покупка жилья  недвижимость      высокий доход  
1     приобретение автомобиля    автомобиль      средний доход  
2               покупка жилья  недвижимость      средний доход  
3  дополнительное образование   образование      высокий доход  
4             сыграть свадьбу       свадьба      высокий доход

In [29]:
df_2 = df[['children','family_status','purpose_group','total_income_group','debt']]

In [30]:
df_2

children     family_status purpose_group total_income_group  debt
0             1   женат / замужем  недвижимость      высокий доход     0
1             1   женат / замужем    автомобиль      средний доход     0
2             0   женат / замужем  недвижимость      средний доход     0
3             3   женат / замужем   образование      высокий доход     0
4             0  гражданский брак       свадьба      высокий доход     0
...         ...               ...           ...                ...   ...
21449         1  гражданский брак  недвижимость      высокий доход     0
21450         0   женат / замужем    автомобиль      высокий доход     0
21451         1  гражданский брак  недвижимость      средний доход     1
21452         3   женат / замужем    автомобиль      высокий доход     1
21453         2   женат / замужем    автомобиль      средний доход     0

[21454 rows x 5 columns]

### Вывод

Необходимо создать категории в канных для того, чтобы облегчить дальнейшую аналитику.
Первое, что необходимо сделать - разбить цели кредита на 4 основные, которые мы определили с помощью стемминага: автомобиль, свадьба, недвижимость, образование. Сделаем это с помощью функции purpose_def. Применим функцию ко всему столбцу 'purpose' с помощью выражения 'df['purpose_group'] = df['purpose'].apply(purpose_def)', формируем новый столбец 'purpose_group'.

Далее необходимо разбить доход клиентов на группы, определяем 3 группы : низкий(до 75000), средний(до 150000), высокий(больше 150000). Уровень среднего дохода формировался на основе median(), которая равна 145017.
С помощью выражения 'df['total_income_group'] = df['total_income'].apply(total_income_def)' применим функцию total_income_def к столбцу total_income для категоризации дохода клиентов, создаем новый столбец total_income_group и помещаем туда новые данные.

<div class="alert alert-block alert-success">
<b>Успех:</b> Категоризация выполнена хорошо.
</div>

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [31]:
df.groupby('children')['debt'].count()

children
0     14091
1      4855
2      2052
3       330
4        41
5         9
20       76
Name: debt, dtype: int64

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 14 columns):
children              21454 non-null int64
days_employed         21454 non-null int64
dob_years             21454 non-null int64
education             21454 non-null object
education_id          21454 non-null int64
family_status         21454 non-null object
family_status_id      21454 non-null int64
gender                21454 non-null object
income_type           21454 non-null object
debt                  21454 non-null int64
total_income          21454 non-null int64
purpose               21454 non-null object
purpose_group         21454 non-null object
total_income_group    21454 non-null object
dtypes: int64(7), object(7)
memory usage: 2.3+ MB


In [33]:
df.loc[df["children"] == 20,"children"] = 0
df.groupby('children')['debt'].count()

children
0    14167
1     4855
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

In [34]:
children_grouped = df.groupby('children').agg({'debt': ['count', 'sum']})
children_grouped

debt      
          count   sum
children             
0         14167  1071
1          4855   445
2          2052   194
3           330    27
4            41     4
5             9     0

In [35]:
sum_debt = df['debt'].sum()
sum_debt

1741

In [36]:
children_grouped['raiting'] = children_grouped['debt']['sum']/children_grouped['debt']['count']
children_grouped

debt         raiting
          count   sum          
children                       
0         14167  1071  0.075598
1          4855   445  0.091658
2          2052   194  0.094542
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000

### Вывод

Определяем наличие детей у клиентов с помощью выражения 'df.groupby('children')['debt'].count()'. В результате замечаем явно выдающееся значение: у 76 клиентов есть по 20 детей, тогда как самое близкое иное количество детей - 5. Чисто 20 считаем артефактом и приравниваем к нулю (наличие в данных таких клиентов 0,003). 

Далее с помощью выражения 'children_grouped = df.groupby('children').agg({'debt': ['count', 'sum']})' формируем таблицу в которой есть столбец количества детей 'children', количество клиентов у которых есть определенное количество детей 'count' и сумма случаев невозврата кредита для данной категории людей 'sum'(1 - есть случае невозврата, 0 - нет случаев невозврата).

После этого необходимо определить вероятность невозврата кредита, основанное на имеющихся данных. Делим число невозвратов кредита на количество пользователей и формируем отдельный столбец 'raiting'. 

Наилучший показатель возврата кредита у людей с 5 детьми, но их и меньше всего. Далее около 9% вероятности просрочки у людей с количеством детей 1,2 и 4. У людей с двумя детьмя - самый лучший показатель возврата для людей у которых есть дети - 8% просрочек. Наилучший показатель возврата кредита у людей, у которых детей нет - около 7%. 

<div class="alert alert-block alert-success">
<b>Успех:</b> Корректно.
</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [37]:
df

children  days_employed  dob_years education  education_id  \
0             1           8437         42    высшее             0   
1             1           4024         36   среднее             1   
2             0           5623         33   среднее             1   
3             3           4124         32   среднее             1   
4             0         340266         53   среднее             1   
...         ...            ...        ...       ...           ...   
21449         1           4529         43   среднее             1   
21450         0         343937         67   среднее             1   
21451         1           2113         38   среднее             1   
21452         3           3112         38   среднее             1   
21453         2           1984         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21449  гражданский брак                 1      F   компаньон     0   
21450   женат / замужем                 0      F   пенсионер     0   
21451  гражданский брак                 1      M   сотрудник     1   
21452   женат / замужем                 0      M   сотрудник     1   
21453   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose purpose_group  \
0            253875                 покупка жилья  недвижимость   
1            112080       приобретение автомобиля    автомобиль   
2            145885                 покупка жилья  недвижимость   
3            267628    дополнительное образование   образование   
4            158616               сыграть свадьбу       свадьба   
...             ...                           ...           ...   
21449        224791             операции с жильем  недвижимость   
21450        155999          сделка с автомобилем    автомобиль   
21451         89672                  недвижимость  недвижимость   
21452        244093  на покупку своего автомобиля    автомобиль   
21453         82047         на покупку автомобиля    автомобиль   

      total_income_group  
0          высокий доход  
1          средний доход  
2          средний доход  
3          высокий доход  
4          высокий доход  
...                  ...  
21449      высокий доход  
21450      высокий доход  
21451      средний доход  
21452      высокий доход  
21453      средний доход  

[21454 rows x 14 columns]

In [38]:
df.groupby('family_status')['debt'].count()

family_status
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
гражданский брак          4151
женат / замужем          12339
Name: debt, dtype: int64

In [39]:
family_status_grouped = df.groupby('family_status').agg({'debt': ['count', 'sum']})
family_status_grouped

debt     
                       count  sum
family_status                    
Не женат / не замужем   2810  274
в разводе               1195   85
вдовец / вдова           959   63
гражданский брак        4151  388
женат / замужем        12339  931

In [40]:
family_status_grouped['raiting'] = family_status_grouped['debt']['sum']/family_status_grouped['debt']['count']
family_status_grouped

debt        raiting
                       count  sum          
family_status                              
Не женат / не замужем   2810  274  0.097509
в разводе               1195   85  0.071130
вдовец / вдова           959   63  0.065693
гражданский брак        4151  388  0.093471
женат / замужем        12339  931  0.075452

### Вывод

Метод аналогичен предыдущему. В данном случае во внимание берем столбец 'family_status' и 'debt'.
В итоге имеем таблицу с категориями семейного статуса и вероятностью просрочки кредита для данной категории.

Итог: 9% невозвратов свойстенно людям не женатым/ не замужним и состоящим в гражданском браке. 7% невозвратов свойственно греппе людей находящихся в разводе и женатым/замужним, наименьший показатель просрочки кредита вдов и вдовцов - 6%.

<div class="alert alert-block alert-success">
<b>Успех:</b> Корректно.
</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [41]:
df.groupby('total_income_group')['debt'].count()

total_income_group
высокий доход     9184
низкий доход      1865
средний доход    10405
Name: debt, dtype: int64

In [42]:
total_income_group_grouped = df.groupby('total_income_group').agg({'debt': ['count', 'sum']})
total_income_group_grouped['raiting'] = total_income_group_grouped['debt']['sum']/total_income_group_grouped['debt']['count']
total_income_group_grouped

debt        raiting
                    count  sum          
total_income_group                      
высокий доход        9184  726  0.079051
низкий доход         1865  136  0.072922
средний доход       10405  879  0.084479

### Вывод

Метод аналогичен предыдущему. Во внимание берем столбцы 'total_income_group' с тремя категорями дохода и 'debt'.

Итог: Для среднего уровня дохода характерен самый высокий показатель просрочки кредита - 8%, далее высокий доход - почти 8% (7.9%), наиболее дисциплинированными плательщика являются люди с низким доходом - 7%. 

<div class="alert alert-block alert-success">
<b>Успех:</b> Корректно.
</div>

- Как разные цели кредита влияют на его возврат в срок?

In [43]:
df.groupby('purpose_group')['debt'].count()

purpose_group
автомобиль       4306
недвижимость    10811
образование      4013
свадьба          2324
Name: debt, dtype: int64

In [44]:
purpose_group_grouped = df.groupby('purpose_group').agg({'debt': ['count', 'sum']})
purpose_group_grouped['raiting'] = purpose_group_grouped['debt']['sum']/purpose_group_grouped['debt']['count']
purpose_group_grouped

debt        raiting
               count  sum          
purpose_group                      
автомобиль      4306  403  0.093590
недвижимость   10811  782  0.072334
образование     4013  370  0.092200
свадьба         2324  186  0.080034

### Вывод

Метод аналогичен предыдущему. Столбцы 'purpose_group' и 'debt'.

Итог: Самый высокий показатель просрочки кредита у людей, которые берут кредит на образовательные цели и покупку автомобиля - 9%, далее кредиты на свадебные цели - 8%, самыми надежными являются цели под операции с недвижимостью - 7%.

<div class="alert alert-block alert-success">
<b>Успех:</b> Все верно.
</div>

### Шаг 4. Общий вывод

Наилучший показатель возврата кредита у людей с 5 детьми, но их и меньше всего. Далее около 9% вероятности просрочки у людей с количеством детей 1,2 и 4. У людей с двумя детьмя - самый лучший показатель возврата для людей у которых есть дети - 8% просрочек. Наилучший показатель возврата кредита у людей, у которых детей нет - около 7%.
Дети - дополнительная статья расходов для семей.

9% невозвратов свойстенно людям не женатым/ не замужним и состоящим в гражданском браке. 7% невозвратов свойственно группе людей находящихся в разводе и женатым/замужним, наименьший показатель просрочки кредита вдов и вдовцов - 6%.
Больше всего доверия к людям семейным - у них 2 источника дохода. Вдовы и вдовцы, видимо, могут расчитывать на накопления бывшего супруга или просто не берут большие суммы.

Для среднего уровня дохода характерен самый высокий показатель просрочки кредита - 8%, далее высокий доход - почти 8% (7.9%), наиболее дисциплинированными плательщика являются люди с низким доходом - 7%. 
Люди с низким уровнем дохода боятся просрочек, тк не смогут выплатить пеню.

Самый высокий показатель просрочки кредита у людей, которые берут кредит на образовательные цели и покупку автомобиля - 9%, далее кредиты на свадебные цели - 8%, самыми надежными являются цели под операции с недвижимостью - 7%.
Образовательные цели и свадьба - кредиты чаще молодых людей, которые пока не прочно стоят на ногах. Просрочки по кредиту на автомобиль скорее всего не очень выгодны, потому и высокие просрочки. Кредиты на недвижимость чаще всего - ипотеки, они долгосрочны, решение людей взвешано, потому эта категория людей самая надежная. 

<div class="alert alert-block alert-success">
<b>Успех:</b> В целом шаг сделан замечательно и очень аккуратно.
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b>  В последнем шаге подразумевается использование метода pivot_table. <br> Однако метод groupby его аналог, поэтому можно считать, что все выполнено корректно. <br> Прошу ознакомиться самостоятельно с <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html"> pivot_table. </a>
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b>
    
Немного советов для **Markdown** стиля, которые могут помочь улучшить визуальность твоих выводов и рассуждений.
    
**BOLD** <br>
*Italics*
    
--- 

# First level heading

--- 
    
## Second level heading
  
---  
Lists:
    
- one 
- two
- three
   
---
    
1. one
2. one two
3. one two three
    
--- 
    
Вывод `переменных`
    
    

    
</div>

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.